### ISSUE TITLE
Add lint for dynamic checks are made against a subtype of the declared type

### ISSUE THREAD

**Issue by plammens** [dart-lang/linter#5088]
Proposal: add a new lint that warns you if you are checking the runtime type of an object against a type that is not a subtype of its declared type (downcasting):
```dart
num foo;
foo is int;  // ok
foo is String;  // warning: according to the static type of foo, foo can never be a String

// also with case statements
if (foo case String string); // warning

// and switch statements/expressions
switch (foo) {
  case String _:  // warning
}
switch (foo) {
  String _ =>  ... // warning
}
```

Use case: to prevent some kinds of type bugs that are not detected at compile time.
For example, in my codebase I made a refactoring that changed the type of a variable `foo` from `T` to `Wrapper<T>`.
I was doing a check like `foo is Completable`, which was now invalid and should have been `foo.data is Completable`.
However, this wasn't detected at compile time, instead at runtime that part would just never run.
